core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java

```diff
@@ -0,0 +1,196 @@
+            List<Consumer<PageReference>> buffers,
+            LocalExchangeMemoryManager memoryManager,
+            long maxBufferedBytes,
+            Function<Page, Page> partitionedPagePreparer,
+            PartitionFunction partitionFunction,
+            int partitionCount,
+            UniformPartitionRebalancer partitionRebalancer)
+    {
+        this.buffers = requireNonNull(buffers, "buffers is null");
+        this.memoryManager = requireNonNull(memoryManager, "memoryManager is null");
+        this.maxBufferedBytes = maxBufferedBytes;
+        this.partitionedPagePreparer = requireNonNull(partitionedPagePreparer, "partitionedPagePreparer is null");
+        this.partitionFunction = requireNonNull(partitionFunction, "partitionFunction is null");
+        this.partitionRebalancer = requireNonNull(partitionRebalancer, "partitionRebalancer is null");
+        this.onPageReleased = forLocalExchangeMemoryManager(memoryManager);
+
+        // Initialize writerAssignments with the buffer size
+        writerAssignments = new IntArrayList[buffers.size()];
+        for (int i = 0; i < writerAssignments.length; i++) {
+            writerAssignments[i] = new IntArrayList();
+        }
+
+        // Initialize partitionAssignments and partitionWriterIndex with the partition limit
+        partitionWriterIndex = new int[partitionCount];
+        partitionAssignments = new IntArrayList[partitionCount];
+        for (int i = 0; i < partitionAssignments.length; i++) {
+            partitionAssignments[i] = new IntArrayList();
+        }
+    }
+
+    @Override
+    public void accept(Page page)
+    {
+        Page partitionPage = partitionedPagePreparer.apply(page);
+
+        // Assign each row to a partition which limits to the partitionCount. If there are more physical partitions than
+        // this artificial partition limit, then it is possible that multiple partitions will get assigned the same
+        // bucket id. Thus, multiple partitions will be scaled together since we track partition physicalWrittenBytes
+        // using the artificial limit (partitionCount). The assignments lists are all expected to cleared by the
+        // previous iterations.
+        for (int position = 0; position < partitionPage.getPositionCount(); position++) {
+            int partition = partitionFunction.getPartition(partitionPage, position);
+            partitionAssignments[partition].add(position);
+        }
+
+        // Scale up writers when current buffer memory utilization is more than 50% of the maximum
+        if (memoryManager.getBufferedBytes() > maxBufferedBytes * 0.5) {
+            partitionRebalancer.rebalancePartitions();
+        }
+
+        // Use Long2IntMap instead of Map<WriterPartitionId, Integer> which helps to save memory in the worst case scenario.
+        // Here first 32 bit of long key contains writerId whereas last 32 bit contains partitionId.
+        Long2IntMap pagePartitionsRowCount = new Long2IntOpenHashMap();
```

Comment:
nit: grammatically it should be `pagePartitionRowCounts`. In other places similar